x-pack/legacy/plugins/apm/public/components/app/TransactionDetails/WaterfallWithSummmary/WaterfallContainer/Waterfall/waterfall_helpers/waterfall_helpers.ts

```diff
@@ -230,65 +280,100 @@ function createGetTransactionById(itemsById: IWaterfallIndex) {
+      case 'error':
+        return getErrorItem(item as APMError);
+    }
+  });
+
+const getChildrenGroupedByParentId = (waterfallItems: IWaterfallItem[]) =>
+  groupBy(waterfallItems, item => (item.parentId ? item.parentId : 'root'));
+
+const getEntryWaterfallTransaction = (
+  entryTransactionId: string,
+  waterfallItems: IWaterfallItem[]
+) =>
+  waterfallItems.find(
+    waterfallItem =>
+      waterfallItem.docType === 'transaction' &&
+      waterfallItem.id === entryTransactionId
+  ) as IWaterfallItemTransaction;
+
 export function getWaterfall(
   { trace, errorsPerTransaction }: TraceAPIResponse,
   entryTransactionId?: Transaction['transaction']['id']
 ): IWaterfall {
   if (isEmpty(trace.items) || !entryTransactionId) {
     return {
-      services: [],
       duration: 0,
       orderedItems: [],
       itemsById: {},
       getTransactionById: () => undefined,
       errorCountByTransactionId: errorsPerTransaction,
+      errorCount: sum(Object.values(errorsPerTransaction)),
       serviceColors: {}
     };
   }
 
-  const waterfallItems = trace.items.map(traceItem => {
-    const docType = traceItem.processor.event;
-    switch (docType) {
-      case 'span':
-        return getSpanItem(traceItem as Span);
-      case 'transaction':
-        return getTransactionItem(
-          traceItem as Transaction,
-          errorsPerTransaction
-        );
-    }
-  });
-
-  const childrenByParentId = groupBy(waterfallItems, item =>
-    item.parentId ? item.parentId : 'root'
+  const waterfallItems: IWaterfallItem[] = getWaterfallItems(
+    trace.items,
+    errorsPerTransaction
   );
-  const entryTransactionItem = waterfallItems.find(
-    waterfallItem =>
-      waterfallItem.docType === 'transaction' &&
-      waterfallItem.id === entryTransactionId
+
+  const childrenByParentId = getChildrenGroupedByParentId(waterfallItems);
+
+  const entryWaterfallTransaction = getEntryWaterfallTransaction(
+    entryTransactionId,
+    waterfallItems
   );
-  const itemsById: IWaterfallIndex = indexBy(waterfallItems, 'id');
-  const orderedItems = entryTransactionItem
-    ? getOrderedWaterfallItems(childrenByParentId, entryTransactionItem)
+
+  const entryTransaction = entryWaterfallTransaction?.transaction;
+
+  const orderedItems = entryWaterfallTransaction
+    ? getOrderedWaterfallItems(childrenByParentId, entryWaterfallTransaction)
     : [];
-  const traceRoot = getTraceRoot(childrenByParentId);
-  const duration = getDuration(orderedItems);
+
+  const traceRoot = getRootTransaction(childrenByParentId);
+  const duration = getTraceDuration(orderedItems);
   const traceRootDuration = traceRoot && traceRoot.transaction.duration.us;
-  const services = getServices(orderedItems);
+  const serviceColors = getServiceColors(orderedItems);
+
+  // the agentMarks should be added direct inside orderedItems, as it doesnt have parent-child relationship
+  orderedItems.push(...getAgentMarks(entryTransaction));
```

Comment:
I'm a little confused about the difference in implementation in this PR and https://github.com/elastic/kibana/pull/53756. Which one are you going for?